### PR TITLE
[Wallet] Allow updating daily limit from backend

### LINKED ISSUE #6731: Configure CICO Limit based on KYC status (for manually KYC'd users)
### Expected Behavior

Please describe the behavior you are expecting

### Current Behavior

What is the current behavior?

### PR DESCRIPTION
**Pull request by gnardini**
### Description

The daily limit of $500 was hard-coded on the app. Now, we can optionally set the `dailyLimit` on Firebase for individual accounts. If there is no limit set on Firebase or the limit set is lower than 500 it will default to 500.

### Tested

Manually.
Screen Shot 2021-02-01 at 18 32 28
Screen Shot 2021-02-01 at 18 18 39

### Related issues

- Fixes #6731

### Backwards compatibility

N/A

### CODE REVIEW

File: packages/mobile/src/firebase/db-rules.json [hunk 1]
```diff
@@ -20,7 +20,15 @@
...
-        ".write": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()"
+        "fcmToken": {
+          ".write": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()"
+        },
+        "language": {
+          ".write": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()"
+        },
+        "dailyLimit": {
+          ".write": false
+        }
```
Comment (jeanregisser) [inline packages/mobile/src/firebase/db-rules.json]: 👍

**Review by jeanregisser — APPROVED** [review]
LGTM 🚢 👍

File: packages/mobile/src/account/reducer.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ export interface State {
   acceptedTerms: boolean
   hasMigratedToNewBip39: boolean
   choseToRestoreAccount: boolean | undefined
+  dailyLimit: number
```
Comment (jeanregisser) [inline packages/mobile/src/account/reducer.ts]:
```suggestion
  dailyLimitCusd: number
```

File: packages/mobile/src/firebase/db-rules.json [hunk 3]
```diff
@@ -20,7 +20,15 @@
...
       "$address": {
         ".read": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()",
-        ".write": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()"
+        "fcmToken": {
+          ".write": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()"
+        },
+        "language": {
+          ".write": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()"
+        },
+        "dailyLimit": {
```
Comment (jeanregisser) [inline packages/mobile/src/firebase/db-rules.json]:
```suggestion
        "dailyLimitCusd": {
```

File: packages/mobile/src/firebase/firebase.ts [hunk 4]
```diff
@@ -247,6 +247,10 @@ export async function notificationsChannel() {
   return simpleReadChannel('notificationsV2')
 }
 
+export async function dailyLimitChannel(address: string) {
+  return simpleReadChannel(`registrations/${address}/dailyLimit`)
```
Comment (jeanregisser) [inline packages/mobile/src/firebase/firebase.ts]:
```suggestion
  return simpleReadChannel(`registrations/${address}/dailyLimitCusd`)
```

File: packages/mobile/src/send/utils.ts [hunk 5]
```diff
@@ -85,8 +86,12 @@ export const getFeeType = (
 }
 
 // exported for tests
-export function dailyAmountRemaining(now: number, recentPayments: PaymentInfo[]) {
-  return DAILY_PAYMENT_LIMIT_CUSD - dailySpent(now, recentPayments)
+export function dailyAmountRemaining(
+  now: number,
+  recentPayments: PaymentInfo[],
+  dailyLimit: number
```
Comment (jeanregisser) [inline packages/mobile/src/send/utils.ts:92]:
```suggestion
  dailyLimitCusd: number
```

**Review by jeanregisser — COMMENTED** [review]
Having second thoughts, I propose we rename the `dailyLimit` field to `dailyLimitCusd` to avoid ambiguity about its unit.
What do you think?

File: packages/mobile/src/firebase/db-rules.json [hunk 6]
```diff
@@ -20,7 +20,15 @@
     "registrations": {
       "$address": {
         ".read": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()",
-        ".write": "auth.uid != null && $address === root.child('users').child(auth.uid).child('address').val()"
+        "fcmToken": {
```
Comment (tarikbellamine) [inline packages/mobile/src/firebase/db-rules.json:23]: What's this for?

File: packages/mobile/src/send/SendAmount.tsx [hunk 7]
```diff
@@ -275,10 +275,10 @@ function SendAmount(props: Props) {
   }, [recipientVerificationStatus, addressValidationType, dollarAmount, getTransactionData, origin])
 
   const onRequest = React.useCallback(() => {
-    if (dollarAmount.isGreaterThan(DAILY_PAYMENT_LIMIT_CUSD)) {
+    if (dollarAmount.isGreaterThan(DEFAULT_DAILY_PAYMENT_LIMIT_CUSD)) {
```
Comment (tarikbellamine) [inline packages/mobile/src/send/SendAmount.tsx:278]: Does this constraint still make sense? With KYC in place, it will become increasingly unlikely that everyone will have the same payment limits. Perhaps the limit should be whatever the user's own limit is rather than assuming the default amount?

File: packages/mobile/src/account/saga.ts [hunk 8]
```diff
@@ -82,6 +84,28 @@ function* initializeAccount() {
...
+export function* watchDailyLimit() {
+  const account = yield call(getConnectedAccount)
+  const channel = yield call(cUsdDailyLimitChannel, account)
+  if (!channel) {
+    return
+  }
+  try {
+    while (true) {
+      const dailyLimit = yield take(channel)
+      if (_.isNumber(dailyLimit)) {
```
Comment (tarikbellamine) [inline packages/mobile/src/account/saga.ts:96]: Why is this necessary? What else can `dailyLimit` be?

File: packages/mobile/src/account/saga.ts [hunk 9]
```diff
@@ -82,6 +84,28 @@ function* initializeAccount() {
   }
 }
 
+export function* watchDailyLimit() {
+  const account = yield call(getConnectedAccount)
```
Comment (tarikbellamine) [inline packages/mobile/src/account/saga.ts]: Do we actually need the account to be "connected"? Seems like this has nothing to do with Geth so probably better to just use `getAccount` right?

**Review by tarikbellamine — COMMENTED** [review]
Nice! Left some minor questions and comments.

File: packages/mobile/src/firebase/db-rules.json (hunk 6 above)
Comment (gnardini) [reply packages/mobile/src/firebase/db-rules.json:23]: So that users can write their token and language but only read their dailyLimit. Otherwise a user would technically be able to change their dailyLimit if they managed to take the firebase key somehow from the binary

File: packages/mobile/src/send/SendAmount.tsx (hunk 7 above)
Comment (gnardini) [reply packages/mobile/src/send/SendAmount.tsx:278]: The other person would have to actually send the funds. If the amount requested is higher than their limit it would probably fail, right? I may be missremembering how the request flow goes exactly.

File: packages/mobile/src/account/saga.ts (hunk 8 above)
Comment (gnardini) [reply packages/mobile/src/account/saga.ts:96]: Just a sanity check. If someone accidentaly writes a string or an object on Firebase this will prevent the app from failing.

File: packages/mobile/src/account/saga.ts (hunk 9 above)
Comment (gnardini) [reply packages/mobile/src/account/saga.ts]: Yup

File: packages/mobile/src/account/saga.ts (hunk 8 above)
Comment (tarikbellamine) [reply packages/mobile/src/account/saga.ts:96]: Can't we restrict the data type of the database to `integer`? Seems like it would be better to stop this as far upstream as possible.

File: packages/mobile/src/firebase/db-rules.json (hunk 6 above)
Comment (tarikbellamine) [reply packages/mobile/src/firebase/db-rules.json:23]: I'm confused...what does the `token` and `language` have to do with `dailyLimit`?

Comment (gnardini) [reply packages/mobile/src/firebase/db-rules.json:23]: All those fields are stored on `registrations/{address}`. Previously, a user could write anything there, so there was a generic `.write` rule. Now, users can write only certain fields there, so there are specific rules for each field.

File: packages/mobile/src/account/saga.ts (hunk 8 above)
Comment (gnardini) [reply packages/mobile/src/account/saga.ts:96]: We will still need a check here to look for empty values since this is an optional field. We can still look into adding a check on Firebase, but I don't see the problem with this particular check here.

Comment (gnardini) [reply packages/mobile/src/account/saga.ts:96]: I added a rule to firebase to restrict to numbers, but those restrictions don't apply to admin users or editing from the console, so there is still room for human error :( https://stackoverflow.com/questions/50658307/validate-firebase-with-rules-when-data-is-written-by-firebase-admin

File: packages/mobile/src/firebase/db-rules.json (hunk 6 above)
Comment (tarikbellamine) [reply packages/mobile/src/firebase/db-rules.json:23]: Ok gotcha. Please include in `Other Changes`!

File: packages/mobile/src/send/SendAmount.tsx (hunk 7 above)
Comment (tarikbellamine) [reply packages/mobile/src/send/SendAmount.tsx:278]: Oh yeah you're right. We skip the `SendAmount` screen for payment requests, which is where the check happens. What you have is probably the best solution for now.

File: packages/mobile/src/account/saga.ts (hunk 8 above)
Comment (tarikbellamine) [reply packages/mobile/src/account/saga.ts:96]: Cool! No problem with this specific check. What bothered me was that it fails silently so I was thinking about how we can prevent it all together.

What do you think about throwing an error if `dailyLimit` is defined but not a number? That way we can have some indication of what went wrong in the logs.

File: packages/mobile/src/firebase/db-rules.json (hunk 6 above)
Comment (gnardini) [reply packages/mobile/src/firebase/db-rules.json:23]: This is not really an other change, this was required to make the feature work, we're adding this property

File: packages/mobile/src/account/saga.ts (hunk 8 above)
Comment (gnardini) [reply packages/mobile/src/account/saga.ts:96]: I can add a log I guess? Throwing an error would be a bit extreme imo, how would you expect the error to be handled? Maybe fire an analytics event?

Comment (tarikbellamine) [reply packages/mobile/src/account/saga.ts:96]: It looks like throwing an error here will get caught and logged by the `catch` block, causing it to log. I agree that displaying an error to the user is probably too much. We probably don't need analytics for this either but your call.

Comment (gnardini) [reply packages/mobile/src/account/saga.ts:96]: Added a log rather than throwing the error so that we're still listening to the Firebase channel and receive updates if they come 😄

**Comment by Lss-Ankit** [comment]
Hi @jeanregisser @tarikbellamine @isabellewei We are blocked to test this task because we don't have enough balance to test, Can you please do the needful to verify this task.
